drift/core/psmc.py

```diff
@@ -209,6 +213,273 @@ def _work_fisher_bias_m(self, mi):
+
+        self.write_fisher_file()
+
+    def q_estimator(self, mi, vec1, vec2, noise=False):
+        """Calculate the quadratic estimator for this mi with data vec1 and vec2"""
+        # if data vector is filled with zeros, return q = 0.0 for this m
+        if np.all(vec1 == 0) and np.all(vec2 == 0):
+            self.qa[mi, :] = 0.0
+
+        # if data vector is empty, return q = 0.0 for this m
+        elif (vec1.shape[0], vec2.shape[0]) == (0, 0):
+            self.qa[mi, :] = 0.0
+
+        else:
+            lside = self.telescope.lmax + 1
+            for bi, bg in self.qa.enumerate(axis=1):
+                for li in range(lside):
+                    lxvec = vec1[:, li]
+                    lyvec = vec2[:, li]
+                    self.qa[mi, bi] += np.sum(
+                        lyvec.conj()
+                        * np.matmul(self.clarray[bi][li].astype(np.complex128), lxvec),
+                        axis=0,
+                    ).astype(np.float64)
+
+        # Calculate q_a for noise power (x0^H N x0 = |x0|^2)
+        if noise:
+            # If calculating crosspower don't include instrumental noise
+            noisemodes = 0.0 if self.crosspower else 1.0
+            evals, evecs = self.kltrans.modes_m(mi)
+            noisemodes = noisemodes + (evals if self.zero_mean else 0.0)
+
+            self.qa.global_slice[mi, -1] = np.sum(
+                (vec1 * vec2.conj()).T.real * noisemodes, axis=-1
+            )
+
+    def recv_send_data(self, data, axis):
+
+        comm = MPI.COMM_WORLD
+        size = comm.Get_size()
+        rank = comm.Get_rank()
+
+        shape = data.shape
+        dtype = data.dtype
+
+        recv_rank = (rank - 1) % size
+        send_rank = (rank + 1) % size
+
+        recv_buffer = np.zeros(shape, dtype=dtype)
+
+        # Need to send in 4GB chunks due to some MPI library.
+        message_size = 4 * 2 ** 30.0
+        dsize = np.prod(shape) * 16.0
+        num_messages = int(np.ceil(dsize / message_size))
+
+        # If dsize =0.0 you get 0 num_messages and it throws error. hack.
+        if num_messages == 0:
+            num_messages = 1
+
+        num, sm, em = mpiutil.split_m(shape[axis], num_messages)
+
+        for i in range(num_messages):
+            slc = [slice(None)] * len(shape)
+            slc[axis] = slice(sm[i], em[i])
+
+            di = np.ascontiguousarray(data[slc], dtype=dtype)
+            bi = np.zeros(di.shape, dtype=dtype)
+
+            # Initiate non-blocking receive
+            request = comm.Irecv(
+                [bi, MPI.DOUBLE_COMPLEX], source=recv_rank, tag=(i * size) + recv_rank
+            )
+            # Initiate send
+            comm.Send([di, MPI.DOUBLE_COMPLEX], dest=send_rank, tag=(i * size) + rank)
+            # Wait for receive
+            request.Wait()
+
+            # Fill recv buffer with messages
+            recv_buffer[slc] = bi
+
+        return recv_buffer
+
+    def split_single(self, m, size):
+
```

Comment:
Docstring would be useful here, to describe in words what this function does

Reply:
Done. Need feedback if it is comprehensible.

Reply:
Why use this routine instead of `mpiutil.partition_list()`?

Reply:
I want every rank to work with one m-mode, with partition list it's not guaranteed that i-th partition will have be length = size, where size is the number of total MPI processes running.